File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/ArtistLibrary.cs
namespace IteratorDesignPattern.Example3;

public class ArtistLibrary : IAggregate<Song>
{
    private readonly Dictionary<string, List<Song>> _artistSongs = new();

    public void AddSong(Song song)
    {
        if (!_artistSongs.ContainsKey(song.Artist))
        {
            _artistSongs[song.Artist] = new List<Song>();
        }
        _artistSongs[song.Artist].Add(song);
    }

    public Dictionary<string, List<Song>> GetArtistSongs() => _artistSongs;

    public IIterator<Song> CreateIterator()
    {
        return new ArtistLibraryIterator(this);
    }

    public IIterator<Song> CreateArtistIterator(string artist)
    {
        return new SpecificArtistIterator(this, artist);
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/ArtistLibrary.java
package behavioral.iterator.example3;

import java.util.*;

public class ArtistLibrary implements IAggregate<Song> {
    private final Map<String, List<Song>> artistSongs = new LinkedHashMap<>();

    public void addSong(Song song) {
        artistSongs.computeIfAbsent(song.getArtist(), k -> new ArrayList<>()).add(song);
    }

    public Map<String, List<Song>> getArtistSongs() { return artistSongs; }

    @Override
    public IIterator<Song> createIterator() { return new ArtistLibraryIterator(this); }

    public IIterator<Song> createArtistIterator(String artist) { return new SpecificArtistIterator(this, artist); }
}
